pkg/agentprofile/agent_profile.go

```diff
@@ -84,29 +89,61 @@ func ProfileToApply(logger logr.Logger, profile *datadoghqv1alpha1.DatadogAgentP
 				logger.Info("conflict with existing profile, skipping", "conflicting profile", profile.Namespace+"/"+profile.Name, "existing profile", existingProfile.String())
 				profileStatus.Conditions = SetDatadogAgentProfileCondition(profileStatus.Conditions, NewDatadogAgentProfileCondition(AppliedConditionType, metav1.ConditionFalse, now, ConflictConditionReason, "Conflict with existing profile"))
 				profileStatus.Applied = metav1.ConditionFalse
-				UpdateProfileStatus(profile, profileStatus, now)
+				UpdateProfileStatus(logger, profile, profileStatus, now)
 				return profileAppliedByNode, fmt.Errorf("conflict with existing profile")
 			} else {
-				nodesThatMatchProfile[node.Name] = true
+				profileLabelValue, labelExists := node.Labels[ProfileLabelKey]
+				if labelExists && profileLabelValue == profile.Name {
+					nodesThatMatchProfile[node.Name] = true
+				} else {
+					nodesThatMatchProfile[node.Name] = false
+					nodesNeedingLabel++
+				}
 				profileStatus.Conditions = SetDatadogAgentProfileCondition(profileStatus.Conditions, NewDatadogAgentProfileCondition(AppliedConditionType, metav1.ConditionTrue, now, AppliedConditionReason, "Profile applied"))
 				profileStatus.Applied = metav1.ConditionTrue
 			}
 		}
 	}
 
-	for node := range nodesThatMatchProfile {
+	useSlowStart := os.Getenv(apicommon.SlowStartEnabled) == "true"
+	numNodesToLabel := 0
+	if useSlowStart {
+		profileStatus.SlowStart = &v1alpha1.SlowStart{}
+		if profile.Status.SlowStart != nil {
+			profileStatus.SlowStart.PodsReady = profile.Status.SlowStart.PodsReady
+			profileStatus.SlowStart.LastTransition = profile.Status.SlowStart.LastTransition
+		}
+		profileStatus.SlowStart.Status = getSlowStartStatus(logger, profile.Status.SlowStart, nodesNeedingLabel)
+
+		if canLabel(logger, profileStatus.SlowStart) {
+			numNodesToLabel = getNumNodesToLabel(logger, profile.Status.SlowStart)
+		}
+	}
+
+	for node, hasCorrectProfileLabel := range nodesThatMatchProfile {
+		if useSlowStart {
+			if hasCorrectProfileLabel {
+				profileStatus.SlowStart.NodesLabeled++
+			} else {
+				if numNodesToLabel <= 0 {
+					continue
+				}
+				numNodesToLabel--
+				profileStatus.SlowStart.NodesLabeled++
+			}
+		}
+
 		profileAppliedByNode[node] = types.NamespacedName{
 			Namespace: profile.Namespace,
 			Name:      profile.Name,
 		}
 	}
```

Comment:
```suggestion
	for node, hasCorrectProfileLabel := range nodesThatMatchProfile {
		if useSlowStart {
			if hasCorrectProfileLabel {
				profileStatus.SlowStart.NodesLabeled++
			} else if numNodesToLabel > 0 {
					profileAppliedByNode[node] = types.NamespacedName{
						Namespace: profile.Namespace,
						Name:      profile.Name,
					}
				numNodesToLabel--
				profileStatus.SlowStart.NodesLabeled++
			}
		} else {
			profileAppliedByNode[node] = types.NamespacedName{
				Namespace: profile.Namespace,
				Name:      profile.Name,
			}
		}


	}
```

would that work to not apply nodes to labels that are already correct?

Reply:
All nodes should be added to `profileAppliedByNode` if a profile applies to them. Otherwise, their profile label will be removed later in the reconcile

Reply:
I find this code a bit challenging to read and well worth refactoring at some point.